OpenRA.Game/Server/Server.cs

```diff
@@ -559,24 +559,32 @@ public void DropClient(Connection toDrop)
 				// Send disconnected order, even if still in the lobby
 				DispatchOrdersToClients(toDrop, 0, new ServerOrder("Disconnected", "").Serialize());
 
-				LobbyInfo.Clients.RemoveAll(c => c.Index == toDrop.PlayerIndex);
-				LobbyInfo.ClientPings.RemoveAll(p => p.Index == toDrop.PlayerIndex);
-
-				// Client was the server admin
-				// TODO: Reassign admin for game in progress via an order
-				if (Dedicated && dropClient.IsAdmin && State == ServerState.WaitingPlayers)
+				// Client was the server admin, but the game isn't over yet.
+				if (Dedicated && dropClient.IsAdmin)
 				{
-					// Remove any bots controlled by the admin
-					LobbyInfo.Clients.RemoveAll(c => c.Bot != null && c.BotControllerClientIndex == toDrop.PlayerIndex);
-
-					var nextAdmin = LobbyInfo.Clients.Where(c1 => c1.Bot == null)
-						.MinByOrDefault(c => c.Index);
+					// Pick the human player who joined right after the previous admin.
+					var nextAdmin = LobbyInfo.Clients.Where(c => c.Bot == null && c != dropClient).MinByOrDefault(c => c.Index);
 
-					if (nextAdmin != null)
+					// Still in the lobby so handle things directly.
+					if (State == ServerState.WaitingPlayers)
 					{
-						nextAdmin.IsAdmin = true;
-						SendMessage("{0} is now the admin.".F(nextAdmin.Name));
+						// Remove the previous admin.
+						LobbyInfo.Clients.RemoveAll(c => c.Index == toDrop.PlayerIndex);
+
+						// Stop pinging when a player drops from the server.
+						LobbyInfo.ClientPings.RemoveAll(p => p.Index == toDrop.PlayerIndex);
+
+						// Remove any bots controlled by the previous admin.
+						LobbyInfo.Clients.RemoveAll(c => c.Bot != null && c.BotControllerClientIndex == toDrop.PlayerIndex);
+
+						if (nextAdmin != null)
+						{
+							nextAdmin.IsAdmin = true;
+							SendMessage("{0} is now the admin.".F(nextAdmin.Name));
+						}
 					}
+					else if (nextAdmin != null)
+						DispatchOrdersToClients(toDrop, 0, Order.NewAdmin(nextAdmin.Index).Serialize());
```

Comment:
Sending this order at frame 0 (i.e. immediate) is asking for desyncs.  All players need to process the frame at the same world tick!  `toDrop.MostRecentFrame` should give the correct result here I think.

Reply:
It crashes the game:

```
Exception of type `System.ArgumentException`: An item with the same key has already been added.
  at System.ThrowHelper.ThrowArgumentException (System.ExceptionResource resource) [0x0000b] in <dca3b561b8ad4f9fb10141d81b39ff45>:0 
  at System.Collections.Generic.Dictionary`2[TKey,TValue].Insert (TKey key, TValue value, System.Boolean add) [0x0008e] in <dca3b561b8ad4f9fb10141d81b39ff45>:0 
  at System.Collections.Generic.Dictionary`2[TKey,TValue].Add (TKey key, TValue value) [0x00000] in <dca3b561b8ad4f9fb10141d81b39ff45>:0 
  at OpenRA.Network.FrameData.AddFrameOrders (System.Int32 clientId, System.Int32 frame, System.Byte[] orders) [0x0000e] in <f2be23e29023499d8549d9e0cebc473c>:0 
  at OpenRA.Network.OrderManager+<TickImmediate>c__AnonStorey0.<>m__0 (System.Int32 clientId, System.Byte[] packet) [0x00077] in <f2be23e29023499d8549d9e0cebc473c>:0 
  at OpenRA.Network.EchoConnection.Receive (System.Action`2[T1,T2] packetFn) [0x00056] in <f2be23e29023499d8549d9e0cebc473c>:0 
  at OpenRA.Network.OrderManager.TickImmediate () [0x000b4] in <f2be23e29023499d8549d9e0cebc473c>:0 
  at OpenRA.Sync+<CheckSyncUnchanged>c__AnonStorey0.<>m__0 () [0x00001] in <f2be23e29023499d8549d9e0cebc473c>:0 
  at OpenRA.Sync.CheckSyncUnchanged[T] (OpenRA.World world, System.Func`1[TResult] fn) [0x00043] in <f2be23e29023499d8549d9e0cebc473c>:0 
  at OpenRA.Sync.CheckSyncUnchanged (OpenRA.World world, System.Action fn) [0x0000e] in <f2be23e29023499d8549d9e0cebc473c>:0 
  at OpenRA.Game.InnerLogicTick (OpenRA.Network.OrderManager orderManager) [0x0010a] in <f2be23e29023499d8549d9e0cebc473c>:0 
  at OpenRA.Game.LogicTick () [0x00050] in <f2be23e29023499d8549d9e0cebc473c>:0 
  at OpenRA.Game.Loop () [0x000d6] in <f2be23e29023499d8549d9e0cebc473c>:0 
  at OpenRA.Game.Run () [0x00042] in <f2be23e29023499d8549d9e0cebc473c>:0 
  at OpenRA.Program.Run (System.String[] args) [0x00011] in <f2be23e29023499d8549d9e0cebc473c>:0 
  at OpenRA.Program.Main (System.String[] args) [0x0004f] in <f2be23e29023499d8549d9e0cebc473c>:0
```